ApiReview/API.Polly.Core/NoDocs/Polly.Retry/OnRetryArguments.cs

```diff
@@ -0,0 +1,14 @@
+// Assembly 'Polly.Core'
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Polly.Retry;
+
+public sealed class OnRetryArguments
```

Comment:
can we change this type to be struct to avoid allocations?

Reply:
This class is never created on the hot-path, only when retry occurs. 

We use classes for all resilience events because they have better binary compatibility, are not boxed and we can still use pooling for them later if needed.

Structs are only used for predicate arguments as those are on hot path and are not ever passed to the telemetry infra (boxing never happens).

Reply:
> This class is never created on the hot-path, only when retry occurs.

yes, but this library should work equally efficient when bad things start to happen and retries become a norm

Reply:
I agree that using structs for everything can simplify the code-base. 

I'll investigate introducing the following API instead of `DiagnosticSource`:

``` csharp
public abstract class TelemetryListener
{
    public abstract void Write<TArgs>(
        ResilienceTelemetrySource source,
        ResilienceEvent resilienceEvent,
        ResilienceContext context,
        TArgs args);

    public abstract void Write<TResult, TArgs>(
        ResilienceTelemetrySource source,
        ResilienceEvent resilienceEvent,
        OutcomeArguments<TResult, TArgs> args);
}
``` 

Binary compatibility should not be an issue (see #1280).

Reply:
All arguments are now structs.